### PR TITLE
[R-package] Remove ability to install precompiled lib_lightgbm (fixes #3320)

### LINKED ISSUE #3320: R README installation instruction hierarchy
The README in `master/R-package` has a title hierarchy that is a bit confusing to me. Here I summarized the part of it that I find hard to follow:

```
Installation
------------
...

### Installing the CRAN package
...

### Installing Precompiled Binaries

...

### Preparation (<- this is a bit confusing to me. 
                      Should it be hierarchy lower and have a header 'Building and installing from source'  
                      or similar that also contains the '### Install'?)

...

#### Windows Preparation

...

#### Windows Toolchain Options

...

#### Mac OS Preparation

...

### Install

...


Examples
--------

...

```

I am excited about recent efforts to distribute this package via CRAN, so maybe this README is WIP anyways and the issue is redunant. Also, since it's quite a long README and contains some information for maintainers and not end users, maybe it's worth considering to move some of this content to vignettes. In any case, thanks for this package.

### PR DESCRIPTION
**Pull request by jameslamb**
This PR has two changes that I couldn't see a clean way to separate, so I've combined them:

* clarify how to install the R package (fixes #3320 )
* remove the ability to use a precompiled `lib_lightgbm.so` with the R package

I think that the use of a precompiled `lib_lightgbm.so` in the R package has been broken for a while: #1730, #2714 . This was a valuable option before we started linking to R when building the library (added in #2901 ) and you could just download `lib_lightgbm.so` from the releases page. Now, it is really difficult to manually build a library that works well with the R package.

I think we should eliminate documentation and code about using a pre-built `lib_lightgbm`, and focus on making customizations you can't get with the CRAN package easier via #2441 (like I recommended in https://github.com/microsoft/LightGBM/issues/3354#issuecomment-688008534)

**current README**

image

**proposal**

image

### CODE REVIEW

File: R-package/README.md [hunk 1]
```diff
@@ -14,63 +19,56 @@
...
-install.packages(
-    pkgs = local_file
-    , type = "binary"
-    , repos = NULL
-)
+remotes::install_url(PKG_URL)
 ```
 
-### Preparation
+### Installing from Source with CMake <a name="install"></a>
```
Comment (jameslamb) [inline R-package/README.md:62]: I've aliased this to `#install` so that old links still work

Comment (lorenzwalthert) [reply R-package/README.md:77]: Wow that's cool. Did not even think this is possible.

Comment (StrikerRUS) [reply R-package/README.md:77]: Indeed, nice hack! But right now contents item isn't working. I believe, removing a space before `<a name ...` will fix it.

image

File: R-package/README.md [hunk 2]
```diff
@@ -5,15 +5,44 @@
 ### Contents
 
 * [Installation](#installation)
+    - [Installing the CRAN package](#installing-the-cran-package)
```
Comment (StrikerRUS) [inline R-package/README.md:8]:
```suggestion
    - [Installing the CRAN Package](#installing-the-cran-package)
```

File: R-package/README.md [hunk 3]
```diff
@@ -5,15 +5,44 @@
...
 * [Known Issues](#known-issues)
 
 Installation
 ------------
 
+For the easiest installation, go to ["Installing the CRAN package"](#installing-the-cran-package).
+
+If you experience any issues with that, try ["Installing from Source with CMake"](#install). This can produce a more efficient version of the library on Windows systems with Visual Studio.
+
+To build a GPU-enabled version of the package, follow the steps in ["Installing a GPU-enabled Build"](#installing-a-gpu-enabled-build)
```
Comment (StrikerRUS) [inline R-package/README.md:26]:
```suggestion
To build a GPU-enabled version of the package, follow the steps in ["Installing a GPU-enabled Build"](#installing-a-gpu-enabled-build).
```

**Comment by lorenzwalthert** [comment]
Looks good. One minor thing: I think the distinction between CPU and GPU packages is very important and hence one could mention that the CRAN package and pre-compiled binaries don't support GPU training (If I got that right). One can figure that out from the header on how to build a GPU accelerated version, but explicit is better than implicit often.

File: R-package/README.md [hunk 4]
```diff
@@ -5,15 +5,44 @@
...
+        objective = "regression"
+        , metric = "l2"
+    )
+    , data = dtrain
+)
+```
+
 ### Installing the CRAN package
 
 As of this writing, `LightGBM`'s R package is not available on CRAN. However, start with `LightGBM` 3.0.0, you can install a released source distribution. This is the same type of package that you'd install from CRAN. It does not require `CMake`, Visual Studio, or anything else outside the CRAN toolchain.
```
Comment (StrikerRUS) [inline R-package/README.md:48]:
```suggestion
As of this writing, `LightGBM`'s R package is not available on CRAN. However, start with `LightGBM` 3.0.0, you can install a released source distribution. This is the same type of package that you'd install from CRAN. It does not require `CMake`, `Visual Studio`, or anything else outside the CRAN toolchain.
```

Comment (StrikerRUS) [reply R-package/README.md:48]: Cannot comment on this line, but we need dot here:

image

File: R-package/README.md [hunk 5]
```diff
@@ -30,47 +59,31 @@ PKG_URL <- "https://github.com/microsoft/LightGBM/releases/download/v3.0.0/light
...
+CXX11=g++-8
 ```
 
-### Preparation
+### Installing from Source with CMake <a name="install"></a>
 
 You need to install git and [CMake](https://cmake.org/) first.
 
-Note: 32-bit (i386) R/Rtools is currently not supported.
+Note: this method is only supported on 64-bit systems. If you need to run LightGBM on 32-bit Windows (i386), follow the instructions in ["Installing the CRAN Package"](#installing-the-cran-package).
```
Comment (StrikerRUS) [inline R-package/README.md:81]: Is it still true?..

File: R-package/README.md [hunk 6]
```diff
@@ -5,15 +5,44 @@
...
 * [Testing](#testing)
-* [Preparing a CRAN Package and Installing It](#preparing-a-cran-package-and-installing-it)
+* [Preparing a CRAN Package](#preparing-a-cran-package)
 * [External Repositories](#external-unofficial-repositories)
 * [Known Issues](#known-issues)
 
 Installation
 ------------
 
+For the easiest installation, go to ["Installing the CRAN package"](#installing-the-cran-package).
```
Comment (StrikerRUS) [inline R-package/README.md:22]: I think we should say here that **the recomended way is CMake + VS** due to performance reasons.

**Review by StrikerRUS — COMMENTED** [review]
Just some minor comments.

File: R-package/README.md (hunk 5 above)
Comment (jameslamb) [reply R-package/README.md:81]: yes, we haven't added support for a 32-bit build with CMake and I don't think we should.

File: R-package/README.md (hunk 6 above)
Comment (jameslamb) [reply R-package/README.md:22]: I disagree with that being "the recommended way". I don't think we should assume that the average user of the R package would prefer efficiency of the library over ease of installation. The CMake path is an order of magnitude more difficult than the CRAN one.

I did note the performance expectations in the next line

> If you experience any issues with that, try ["Installing from Source with CMake"](#install). This can produce a more efficient version of the library on Windows systems with Visual Studio.

File: R-package/README.md (hunk 1 above)
Comment (jameslamb) [reply R-package/README.md:77]: the spacing for HTML tags is irrelevant, except for very special ones like `<pre>`. The issue here was that I got the anchor wrong in the table of contents: https://github.com/microsoft/LightGBM/pull/3360/commits/df5bf3f6b193b68a9e83dc16b5e31341f142a355

this is working now

image

**Comment by jameslamb** [comment]
> Looks good overall. One minor thing: I think the distinction between CPU and GPU packages is very important and hence one could mention that the CRAN package and pre-compiled binaries don't support GPU training (If I got that right). One can figure that out from the header on how to build a GPU accelerated version, but explicit is better than implicit often.

thanks @lorenzwalthert . I disagree with this. As you mentioned in #3320, the README is already hard to follow and has a lot of information. I want to make this as simple as possible.

I think anyone arriving at this page looking to build for the GPU will be drawn to "Installing a GPU-enabled Build" in the list of options.

image

**Review by StrikerRUS — APPROVED** [review]

File: R-package/README.md (hunk 6 above)
Comment (StrikerRUS) [reply R-package/README.md:22]: cc @guolinke and @Laurae2

Comment (jameslamb) [reply R-package/README.md:22]: I won't merge this until this thread is answered. Gently ping @guolinke

**Comment by jameslamb** [comment]
I just merged `master` into this. This is ready for review and does not need to wait for #3338

File: R-package/README.md (hunk 6 above)
Comment (guolinke) [reply R-package/README.md:22]: sorry for missing this. For cran and R, I think easy-to-install is preferred by users. So we can add a note about cmake+vs is more efficiency in windows.

**Review by guolinke — APPROVED** [review]

File: R-package/README.md (hunk 6 above)
Comment (jameslamb) [reply R-package/README.md:22]: ok great, thanks! Then I think this is ok as of I've written it, with the note on the the choice of cmake + VS for a more efficient version

**Comment by jameslamb** [comment]
I just merged `master` into this. Since we've agreed on https://github.com/microsoft/LightGBM/pull/3360#discussion_r484522114, I'll merge this one it builds.
